tests-e2e/gappy_state_test.go

```diff
@@ -0,0 +1,62 @@
+package syncv3_test
+
+import (
+	"github.com/matrix-org/sliding-sync/sync3"
+	"github.com/matrix-org/sliding-sync/testutils/m"
+	"testing"
+)
+
+func TestGappyState(t *testing.T) {
+	t.Log("Alice registers on the homeserver.")
+	alice := registerNewUser(t)
+
+	t.Log("Alice creates a room")
+	roomID := alice.CreateRoom(t, map[string]interface{}{"preset": "public_chat"})
+
+	t.Log("Alice sends a message into that room")
+	alice.SendEventSynced(t, roomID, Event{
+		Type: "m.room.message",
+		Content: map[string]interface{}{
+			"msgtype": "m.text",
+			"body":    "Hello, world!",
+		},
+	})
+
+	t.Log("Alice logs out of her first device.")
+	alice.MustDoFunc(t, "POST", []string{"_matrix", "client", "v3", "logout"})
+
+	t.Log("Alice logs in again on her second device.")
+	alice.Login(t, "password", "device2")
+
+	t.Log("Alice sets two pieces of room state while the proxy isn't polling.")
+	nameContent := map[string]interface{}{"name": "potato"}
+	nameID := alice.SetState(t, roomID, "m.room.name", "", nameContent)
+	powerLevelState := map[string]interface{}{
+		"users":          map[string]int{alice.UserID: 100},
+		"events_default": 10,
+	}
+	powerLevelID := alice.SetState(t, roomID, "m.room.power_levels", "", powerLevelState)
+
+	t.Log("Alice logs out of her second device.")
+	alice.MustDoFunc(t, "POST", []string{"_matrix", "client", "v3", "logout"})
+
+	t.Log("Alice logs in again on her third device.")
```

Comment:
This feels unecessary. The point of using >1 device is to refresh the token. It should be fine to just log out first -> send lots of events with 2nd device -> do a SS request. You need to send > timeline limit events in the room to force a gappy sync.

Reply:
Thanks, that makes more sense. To check: I'm trying to hit this log line

https://github.com/matrix-org/sliding-sync/blob/1f3f14f30c98cae687145a6cd0fabc097d571c9c/state/accumulator.go#L155-L159

which gets called when we "ignore" a state block in a sync response?

Reply:
Yes